libbpf-tools/biopattern.bpf.c

```diff
@@ -17,12 +18,24 @@ struct {
 } counters SEC(".maps");
 
 SEC("tracepoint/block/block_rq_complete")
-int handle__block_rq_complete(struct trace_event_raw_block_rq_complete *ctx)
+int handle__block_rq_complete(void *args)
 {
-	sector_t sector = ctx->sector;
 	struct counter *counterp, zero = {};
-	u32 nr_sector = ctx->nr_sector;
-	u32 dev = ctx->dev;
+	sector_t sector;
+	u32 nr_sector;
+	u32 dev;
+
+	if (has_block_rq_completion()) {
+		struct trace_event_raw_block_rq_completion___x *ctx = args;
+		sector = BPF_CORE_READ(ctx, sector);
+		nr_sector = BPF_CORE_READ(ctx, nr_sector);
+		dev = BPF_CORE_READ(ctx, dev);
+	} else {
+		struct trace_event_raw_block_rq_complete *ctx = args;
```

Comment:
What if the vmlinux.h is updated with newer kernel and trace_event_raw_block_rq_complete type is not available?
Should we define this type in the header file as well?

Reply:
Yes, that will greatly improve the portability.

But I would like to defer that to another PR, and include other type definitions (for struct task_struct etc).

Reply:
Okay, sounds good. We can have follow-up patch once vmlinux.h is updated in the future.